Travis releases

Kanaye opened the pull request:
With the graphql api it's possible to fetch everything we need.
The api is still early access/alpha and so has some bugs I needed to work arround and some things that don't work as smooth as I hoped. But that's okay.
Because it's an alpha version some things *could* go wrong with a release, but nothing I shouldn't be able to fix within a few minutes. So I guess that's fine.
GraphQL is nicer than normal REST APIs (previously I hit the limit of github api calls) and now it's maybe a couple dozen of calls instead of a couple hundred.
I would like it to have more options to query or to set vars within an query (SQL subselect-ish) but I guess it's a start.

Detection before it tries to publish anything:
- checks if ``GIT_KEY``, ``GIT_USER`` and ``GIT_EMAIL`` are set in the env.
- checks if the package.jsons version property in the project root has been changed in the last commit.
- checks if there is yet a release on github with the same version (prevents errorss if somehow a travis build is triggert twice for one commit).

Github releases work fine. 
It tags, commits and then pushes to the repo.
After that it fetches the issues, prs, build a release log and publishes it with the github release (currently via the REST api the GraphQL API can't create releases (yet)).

With the last commit it also publishes/updates the npm package between pushing the tag&commit and publishing the release.
It's not tested, but it just needs the ``NPM_TOKEN`` env var set and after switching the processes cwd to the dist/npm dir the [npm-utils](https://www.npmjs.com/package/npm-utils) node module does the rest.

[Here](https://github.com/Kanaye/jsblocks/releases/tag/0.3.5-alpha1) the test github release on my jsblocks fork.
And [here](https://travis-ci.org/Kanaye/jsblocks) the tavis log of the build.

The release will be availble on:
- github
- bower
- npm
- jsdelivr (because there bot also fetches github)

Tell me if I missed something.

This add 3 new dev dependencies:
1. node-fetch: just a pure-js dependency, I just like fetch more than the node-api and because it's just a small dependencie I took it
2. npm-utils: a wrapper for the npm cli that also handles auth via env vars.
3. nodegit: that's the heavy one, it requires a  c++ 11 compiler on exotic node versions (for lts and current versions it just downloads binaries on the supported platforms).
It also required me to let travis fetch a current c++ 11 stdlib, makes travis builds a little slower.
If you whish we can move that dependencies to an other package.json and only install them if we need them.
I also updated the nodejs version travis uses to tha last LTS, I think the last LTS is the version we should focus on and also some of the dependencies required a current release.

I also created a JSBlocks-CI [github](https://github.com/JSBlocks-CI) and [npm](https://www.npmjs.com/~jsblocks-ci) user. 
**[edit:]** 
These two have currently the email address: jsblocks-ci@kanaye.net running on my mailserver if you wish we can change that to some other address (e.g. a @jsblocks.com address) otherwise I'm fine with leaving them running on my mail-server.
**[/edit]**
Once these two have access and I added the env-vars to travis we should be ready to go.

Tell me what you think.

Kanaye commented:
Things I forgot to mention earlier:
- Travis commits with the message:  `` [ci skip] Build Version {{version}}``
   The ``[ci skip]`` ensures that current build doesn't get interrupted because travis tests the new commit (wouldn't happen with our current travis config, but also that commit doesn't change the source files so there is nothing new to test).
- If the version contains alpha, beta or rc it will be flagged as a prerelease version on github.
- Because I'm using the travis ``after_success`` hook  travis won't show the tests as failed if the publishing fails. So we should be suspicious if we don't get a release notification from github after a while.
- closed issues and merged PRs are included in the release log if they have a ``ClosedEvent`` or ``ReferencedEvent`` on issues or a ``MergedEvent`` in their timeline with a commit after the last release. Because I can only fetch the first 100 events for a issue/pr in a single request (and I don't feel like making more requests for that purpose is necessary) a pr/issue would not get mentioned in the release if it has been closed/merged after it already had 100 events. Shouldn't happen right now because we don't have that many comments/subscriptions/* on a single issue/pr and if we reach that point I hope the GraphQL API will have more filter options.

astoilkov commented:
Wow. This is so much quality work. Thank you for all your contributions.

Regarding the users: No problem. Leave it using your email if that's fine with you. We could transfer in the future.

Regarding the new dependencies:
  - I don't like big dependencies like nodegit but if nothing else gets the job done I always feel confident that's the right decision. Leave it in `devDependencies`
  - node-fetch - perfect that you found something smaller and minimalistic as a dependency. Btw I use [got](https://github.com/sindresorhus/got) for making requests. You could check it out.

Do you think you want to make more changes or we can merge this?

Kanaye commented:
I also don't like big dependencies, but it's the only thing I found that could handle authentication with keys without major problems and add the changed files. Otherwise I would have used something that e.g. just wraps the git-cli.

I took a look at got some time ago but haven't used it yet, so I also kind of forgot it.

At least now I have now changes in mind, 
so I think we can merge it and I'll add the env-vars to travis when I'm back

Kanaye commented:
I just added the GIT_* env vars for @jsblocks-ci and the NPM_TOKEN for the [npm-user](https://www.npmjs.com/%7Ejsblocks-ci) so the only thing left before it'll work is to give them the access rights.

If we would push a version change in the package.json right now it would fail with printing the error from nodegit that it can't push.

Also the publish task will always try to push to the git repository first and than publish anything else for the case something got pushed after travis started the build, in that case the git push would just fail and abort everything instead of leaving a release mess to clean up.

Regarding the email: I'm fine with that, one more mail address won't hurt my servers ;)

I think the last thing I wanted to do before the next (and probably for now last)  release with jsvalue was to update it to make sure the functions I moved/reimplemented in jscore don't cause any trouble for users that currently depend on it.
For some reason I currently have a dirty head on my local jsvalue repo and I want to make sure I've got every thing right before updating. Something I'll look over within probably the next 2 days.

I'll ping you here when that's done.

Kanaye commented:
While I have some still have some changes on my local jsvalue repository. 
Nothing of these changes is major (or done) enough.
I pushed and tested the changes.
So from my point everything is ready for the next release 🎉

astoilkov commented:
So let's do it. Anything I can help with?

Kanaye commented:
Yes, you need to invite the to above mentioned users to this repo and npm.
After that's done and I accepted the invites,
it's up to you who of us should commit the version change in the package.json.

astoilkov commented:
It's done for the repo but in the npm it says JSBlocks-CI and jsblocks-ci@kanaye.net users don't exist. What is the username in npm that I should invite?

Kanaye commented:
the email address is right but afaik npm doesn't support uppercase in usernames so you could try jsblocks-ci.
https://www.npmjs.com/~jsblocks-ci is the url to the npm profile.

Kanaye commented:
Just tested: lowercase jsblocks-ci works on npm :)
Also accepted the github invite on the ci user.

astoilkov commented:
done

Kanaye commented:
Nice. 
If you wish you can just push the package.json with a version change to 0.3.5 (or edit it on github) I'll stay around for the case I missed something.

astoilkov commented:
done

Kanaye commented:
And we have a new release :tada:
Looks like everything worked fine

astoilkov commented:
Everything looks amazing. I really like the release notes. The formatting is awesome.

Kanaye commented:
Thanks.
Release notes in that style (closed issues, commits) are something I'm always looking for when I upgrade some libs.
So I liked the Idea of automating it.
The formatting is adapted from the nodejs.
Glad you like it :+1:
